ext/background/api.js

```diff
@@ -5,15 +5,12 @@ const manifest = chrome.runtime.getManifest();
 // 30 minutes
 const sessionExpiryThreshold = 30 * 60 * 1000;
 
-let _sessionCache = {};
-let _invalidCookieCache = {};
-let init = refreshSessions(true);
+// Ensure that sessions are initialized before API calls are made
+let initSessions = refreshSessions();
```

Comment:
My understanding is that we still need to get rid of this `initSessions` global promise variable. The background page is not persistent, and so all of the global variables may be re-initialized any number of times at any frequency, depending on how frequently the background page gets re-initialized. Is it your intention that `refreshSessions()` could potentially be called very frequently here?

If not, I think that we may need to change how we store the state for whether sessions are initialized or not. Perhaps `refreshSessions` could be called by:
1. The `cookies.onChanged` listener.
1. The `request` method, if no sessions are initialized in the local storage yet.
1. The `request` method, if needed due to an error code.
1. Is there a `chrome.sessions.onStart` or `chrome.extension.onInstall` listener that we can use to kick off the initial `refreshSessions` call instead of doing it here in the global scope?

Reply:
Having this init variable helps ensure we aren't dropping the first requests made when the background page pops up.

`refreshSessions()` without using the `force` arg will just do a local scan of cookies and timestamps. It will only make a login call if something is expired or expiring within 30 minutes. I think this is good at the time that the background page is created as it ensures that no requests are dropped if the background page is regularly killed (which it appears it is). Once we are in the `request()` method it's too late to instantiate the session cache and we will have a failed request. Some sort of refresh+retry logic would need to be introduced.

Not using `chrome.storage` and `refreshSessions(true)` was causing the issue. Not using `chrome.storage` caused the session state to be wiped when background pages were restarted. Passing the `force` arg would also cause session state to be wiped upon the instantiation of the background page. Because session state isn't being lost when the background page is killed, we should be good now I think.

Reply:
OK, got it! That makes sense to me now that you've described how it should work, thanks for walking me through it 🙂